codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -375,13 +382,58 @@ public void write(int allowedBytes) {
                         writePromise = ctx.newPromise();
                         writePromise.addListener(this);
                     }
+                    if (toWrite instanceof SlicedByteBuf && data instanceof CompositeByteBuf) {
+                        // If we're writing a subset of a composite buffer then we want to release
+                        // any underlying buffers that have been consumed. CompositeByteBuf only releases
+                        // underlying buffers on write if all of its data has been consumed and its refCnt becomes
+                        // 0.
+                        final CompositeByteBuf toFree = (CompositeByteBuf) data;
+                        writePromise.addListener(new ChannelFutureListener() {
+                            @Override
+                            public void operationComplete(ChannelFuture future) throws Exception {
+                                toFree.discardReadComponents();
+                            }
+                        });
+                    }
                     frameWriter().writeData(ctx, stream.id(), toWrite, writeablePadding,
-                        size == bytesWritten && endOfStream, writePromise);
+                            size == bytesWritten && endOfStream, writePromise);
                 } while (size != bytesWritten && allowedBytes > bytesWritten);
             } finally {
                 size -= bytesWritten;
             }
         }
+
+        @Override
+        public boolean merge(Http2RemoteFlowController.FlowControlled next) {
+            if (FlowControlledData.class != next.getClass()) {
+                return false;
+            }
+            FlowControlledData nextData = (FlowControlledData) next;
+            // Given that we're merging data into a frame it doesn't really make sense to accumulate padding.
+            padding = Math.max(nextData.padding, padding);
+            endOfStream = nextData.endOfStream;
+            final CompositeByteBuf compositeByteBuf;
+            if (data instanceof CompositeByteBuf) {
+                compositeByteBuf = (CompositeByteBuf) data;
+            } else {
+                compositeByteBuf = ctx.alloc().compositeBuffer(Integer.MAX_VALUE);
+                compositeByteBuf.addComponent(data);
+                compositeByteBuf.writerIndex(data.readableBytes());
+                data = compositeByteBuf;
+            }
+            compositeByteBuf.addComponent(nextData.data);
+            compositeByteBuf.writerIndex(compositeByteBuf.writerIndex() + nextData.data.readableBytes());
+            size = data.readableBytes() + padding;
+            if (!nextData.promise.isVoid()) {
+                // Replace current promise if void otherwise chain them.
+                if (this.promise.isVoid()) {
+                    this.promise = nextData.promise;
+                } else {
+                    this.promise.addListener(new ChannelPromiseNotifier(nextData.promise));
```

Comment:
Rather than creating a ChanenlPromiseNotifier for each merged promise, would it make sense to create a similar class that permits appending promises to it?